packages/sqflite/tizen/src/database_manager.cc

```diff
@@ -0,0 +1,322 @@
+#include "database_manager.h"
+
+#include <flutter/standard_method_codec.h>
+#include <sqlite3.h>
+
+#include <list>
+#include <variant>
+
+#include "errors.h"
+#include "log.h"
+
+namespace sqflite_database {
+DatabaseManager::~DatabaseManager() {
+  for (auto &&stmt : stmt_chache_) {
+    FinalizeStmt(stmt.second);
+    stmt.second = nullptr;
+  }
+  if (database_ != nullptr) {
+    Close();
+  }
+}
+
+void DatabaseManager::ThrowCurrentDatabaseError() {
+  throw sqflite_errors::DatabaseError(GetErrorCode(), GetErrorMsg());
+}
+
+void DatabaseManager::Init() {
+  LOG_DEBUG("initializing database");
+  int result_code = SQLITE_OK;
+  result_code = sqlite3_shutdown();
+  if (result_code != SQLITE_OK) {
+    ThrowCurrentDatabaseError();
+  };
+  result_code = sqlite3_config(SQLITE_CONFIG_URI, 1);
+  if (result_code != SQLITE_OK) {
+    ThrowCurrentDatabaseError();
+  }
+  result_code = sqlite3_initialize();
+  if (result_code != SQLITE_OK) {
+    ThrowCurrentDatabaseError();
+  }
+}
+
+void DatabaseManager::Open() {
+  Init();
+  LOG_DEBUG("opening/creating read write database");
+  int result_code =
+      sqlite3_open_v2(path_.c_str(), &database_,
+                      SQLITE_OPEN_READWRITE | SQLITE_OPEN_CREATE, NULL);
+  if (result_code != SQLITE_OK) {
+    ThrowCurrentDatabaseError();
+  }
+  result_code = sqlite3_busy_timeout(database_, kBusyTimeoutMs);
+  if (result_code != SQLITE_OK) {
+    sqlite3_close(database_);
+    ThrowCurrentDatabaseError();
+  }
+}
+
+void DatabaseManager::OpenReadOnly() {
+  Init();
+  LOG_DEBUG("open read only database");
+  int result_code =
+      sqlite3_open_v2(path_.c_str(), &database_, SQLITE_OPEN_READONLY, NULL);
+  if (result_code != SQLITE_OK) {
+    ThrowCurrentDatabaseError();
+  }
+  result_code = sqlite3_busy_timeout(database_, kBusyTimeoutMs);
+  if (result_code != SQLITE_OK) {
+    sqlite3_close(database_);
+    ThrowCurrentDatabaseError();
+  }
+}
+const char *DatabaseManager::GetErrorMsg() { return sqlite3_errmsg(database_); }
+
+int DatabaseManager::GetErrorCode() {
+  return sqlite3_extended_errcode(database_);
+}
+
+void DatabaseManager::Close() {
+  LOG_DEBUG("closing database");
+  int result_code = sqlite3_close(database_);
+  if (result_code) {
+    ThrowCurrentDatabaseError();
+  }
+}
+
+void DatabaseManager::BindStmtParams(DatabaseManager::Statement stmt,
+                                     SQLParameters params) {
+  int err = SQLITE_OK;
+  const int params_length = params.size();
+  LOG_DEBUG("received %d params to execute sql", params_length);
+  for (int i = 0; i < params_length; i++) {
+    auto idx = i + 1;
+    auto param = params[i];
+    switch (param.index()) {
```

Comment:
What is each case based on? What rules are the parameters created by?
I haven't found any rationale for this implementation.

Reply:
Hey @bbrto21, each param is a `flutter::EncodableValue`, which is a `std::variant`. We're switching based on each variant alternative

Based on each variant, we need to apply a different strategy to bind parameters to the SQL statement in Sqlite3

Reply:
Thanks for the reply, but I wonder where the constants in each case below come from.

``` c++
case 0:
case 1:
case 2:
...
```

Reply:
So, whenever you use `std::variant`, you can determine which was the assigned type of the variable by accessing `index()`:
https://en.cppreference.com/w/cpp/utility/variant/index

For `flutter::EncodableValue` these are the possible types:

```
std::variant<
    std::monostate,  // 0
    bool,  // 1
    int32_t,  // 2
    int64_t,  // 3
    double,  // 4
    std::string,  // 5
    std::vector<uint8_t>,  // 6
    std::vector<int32_t>,  // 7
    std::vector<int64_t>,  // 8
    std::vector<double>,  // 9
    EncodableList,  // 10
    EncodableMap,  // 11
    CustomEncodableValue>   // 12
```

Reply:
@rmanganiello Thanks for checking!